### ISSUE TITLE
Duplicated forum shortcodes?

### ISSUE THREAD

**Issue by rica-carv** [e107inc/e107#2060]
Why there's duplicated forum shortcodes in forum_shortcodes.php & viewforum_shortcodes.php?
And why some have different output of the same shortcode on the other file?
Is it intended, or not?

**Comment by Moc** [comment]
These are probably intended. Can you give us an example?

**Comment by rica-carv** [comment]
@Moc Here goes:
sc_avatar($opts)
sc_forumtitle()
sc_iconkey()
sc_lastpost() 
sc_lastpostdate()
sc_lastpostitle()
sc_lastpostuser()
sc_newflag()
sc_perms()
sc_replies()
sc_repliesx()
sc_replytitle()
sc_search()
sc_startertitle()
sc_threadtitle()

I wonder, if some of these shortcodes do the same, wouldn't be more "clean" to not have them duplicated?

**Comment by Moc** [comment]
Hmm, you are right. Something strange is happening here. 

- In version 2.1.1, the `e107_plugins/forum/shortcodes/batch` folder only contained two files: view and post shortcodes. 
- In version 2.1.2, two additional files are added. This was actually done by you @rica-carv in this commit: https://github.com/e107inc/e107/commit/d4297af1bec4f5805bbb9a8d410a1df87b76fa4c

This needs cleaning up. Labelling and milestoned for next release. 

@CaMer0n Can you take a peak to see what is the best way to solve this? (ie. what can be deleted and what should remain). I am asking this as you made major changes to the new files in this commit: https://github.com/e107inc/e107/commit/1f3cb0d47bcd0fcbeb210c51d11972168e535248

**Comment by rica-carv** [comment]
@Moc Yes, i did create the files, when "cleaning up" the vars from the forum main files.
But i simply transfer the variables code to the sc_functions wich had to retain the same name of the original ones...
Anyway, i agree that the code might need a clean up, so that's why i post this question. It's not a real bug, since the forum still works... At least, for me... 😃

**Comment by rica-carv** [comment]
Some more investigation:
Isn't the sc_search shortcode in forums, exactly the same as the core search shortcode?
After all, both forms call the search.php file.....

**Comment by CaMer0n** [comment]
They could probably be merged into a single forum_shortcodes.php class.
Every shortcode would need to be tested alongside the other (including any parm options) to see where the output varies. 
When it varies, we'd need to rename the shortcode, and then search/replace that shortcode in the template with the new name. ie. `str_replace("{OLDSHORTCODE", "{NEWSHORTCODE", $template) ` to maintain backward compatibility. 

It may be easier to leave it as it is, unless there is a conflict, in which case we just rename some shortcodes and forget about the merging part.

**Comment by rica-carv** [comment]
@CaMer0n Well, as stated before, it might be some conflict with the {SEARCH} shortcode, since the ones that's in forum plugin will take precedence over the core one, when they both do exactly the same:
[Core {SEARCH}](https://github.com/e107inc/e107/blob/afbe20614b46805fc8fb169f5dd0a200bf600062/e107_core/shortcodes/single/search.sc):
```
global $sql,$sysprefs,$SEARCH_SHORTCODE;

include_lan(e_PLUGIN."search_menu/languages/".e_LANGUAGE.".php");
$text = "";
if (!isset($SEARCH_SHORTCODE)) 
{
	if (file_exists(THEME."search_template.php")) 
	{
			include(THEME."search_template.php");
	} 
	else 
	{
	  		include(e_CORE."templates/search_template.php");
	}
}
	$ref['all'] = 'all';
	$ref['news'] = '0';
	$ref['comments'] = 1;
	$ref['users'] = 2;
	$ref['downloads'] = 3;
	$ref['pages'] = 4;

//	$search_prefs = $sysprefs -> getArray('search_prefs');
	
	$search_prefs = e107::getConfig('search')->getPref();


	if(is_array($search_prefs['plug_handlers']))  
	{
	    foreach ($search_prefs['plug_handlers'] as $plug_dir => $active) 
	    {
			if (is_readable(e_PLUGIN.$plug_dir."/e_search.php")) 
			{
				$ref[$plug_dir] = $plug_dir;
			}
		}
	}
	if($ref[$parm]!= ''){
        $page = $ref[$parm];
    }elseif($parm='all' && $ref[e_PAGE] != ''){
    	$page = $ref[e_PAGE];
	}else{
    	$page = 'all';
	}

	$text .= "<form method='get' action='".e_HTTP."search.php'><div>\n";
	$text .= "<input type='hidden' name='t' value='$page' />\n";
	$text .= "<input type='hidden' name='r' value='0' />\n";

	if (defined('SEARCH_SHORTCODE_REF') && SEARCH_SHORTCODE_REF != '') {
	  	$text .= "<input type='hidden' name='ref' value=\"".SEARCH_SHORTCODE_REF."\" />\n";
	}

	$text .= $SEARCH_SHORTCODE;

	$text .="\n</div></form>";

return $text;
```
[Forum {SEARCH}](https://github.com/e107inc/e107/blob/b2cd818a900e80f02eb5cf0d5919ff025566096b/e107_plugins/forum/shortcodes/batch/forum_shortcodes.php#L166):
```
	function sc_search()
	{
		 if(!$srchIcon = e107::getParser()->toGlyph('fa-search'))
		{
			$srchIcon = LAN_SEARCH;
		}
		// String candidate for USERLIST wrapper
		return "
		<form method='get' class='form-inline input-append' action='".e_BASE."search.php'>
		<div class='input-group'>
		<input type='hidden' name='r' value='0' />
		<input type='hidden' name='t' value='forum' />
		<input type='hidden' name='forum' value='all' />
		<input class='tbox form-control' type='text' name='q' size='20' value='' maxlength='50' />
		<span class='input-group-btn'>
		<button class='btn btn-default button' type='submit' name='s' value='search' />".$srchIcon."</button>
		</span>
		</div>
		</form>\n";
	}
```
[Viewforum {SEARCH}](https://github.com/e107inc/e107/blob/fe899855ab9f057185bdb8a84225be3e1016f9c2/e107_plugins/forum/shortcodes/batch/viewforum_shortcodes.php#L273):
```
		function sc_search()
		{
			return "
			<form method='get' class='form-inline input-append' action='" . e_BASE . "search.php'>
			<p>
			<input class='tbox' type='text' name='q' size='20' value='' maxlength='50' />
			<button class='btn btn-default button' type='submit' name='s' >" . LAN_SEARCH . "</button>
			<input type='hidden' name='r' value='0' />
			<input type='hidden' name='ref' value='forum' />
			</p>
			</form>";
		}
```

I can't really see any difference in the HTML, since all the three shortcodes form end up calling the search.php file.
Now if the forum search shortcodes could only narrow the search inside the forum itself, then they could make the difference....

**Comment by Jimmi08** [comment]
Not study code example above, but forum search shortcode search only forum, nothing more. Let it be this way please.

**Comment by Moc** [comment]
So it appears I'm running into issues because of this:
the shortcode `REPLIESX`  (` function sc_repliesx()`) is defined both in forum_shortcodes.php and viewforum_shortcodes.php, and they differ. 

This overly complicates the forum plugin once again. We need to clean this up.
